db/migrations/20201214175739_jobs_table.sql

```diff
@@ -0,0 +1,356 @@
+-- migrate:up
+create table jobs (
+  id binary(16) not null default (uuid_to_bin(uuid(), 1)),
+  user_id binary(16) not null,
+  system_id binary(16) not null,
+  definition JSON not null,
+  status enum('created', 'queued', 'complete', 'error') default 'created',
+  created_at timestamp not null default current_timestamp,
+  modified_at timestamp not null default current_timestamp on update current_timestamp,
+  primary key (id),
+  key jobs_user_id_key (user_id),
+  foreign key (user_id)
+  references users(id)
+  on delete cascade on update restrict
+) engine=innodb row_format=compressed;
+
+create table job_data (
+  id binary(16) not null default (uuid_to_bin(uuid(), 1)),
+  job_id binary(16) not null,
+  schema_path varchar(128) not null,
+  type varchar(32) not null,
+  present boolean not null default false,
+  format varchar(64),
+  -- query this to check instead of data which may need to fetch from disk
+  filename varchar(128),
+  -- store data in table for now, future iterations may store data in an object
+  -- store like S3 and instead link to it here
+  data longblob,
+  created_at timestamp not null default current_timestamp,
+  modified_at timestamp not null default current_timestamp on update current_timestamp,
+  primary key (id),
+  key job_data_id_key (job_id),
+  foreign key (job_id)
+  references jobs(id)
+  on delete cascade on update restrict
+) engine=innodb row_format=dynamic;
+
+
+create definer = 'select_objects'@'localhost'
+  function check_users_job (auth0id varchar(32), jobid char(36))
+    returns boolean
+    comment 'Check if a job exists and belongs to the user'
+    reads sql data sql security definer
+  begin
+    return exists(select 1 from jobs where id = uuid_to_bin(jobid, 1)
+                                       and user_id = get_user_binid(auth0id));
+  end;
+
+grant select on jobs to 'select_objects'@'localhost';
+grant execute on function `check_users_job` to 'select_objects'@'localhost';
+
+-- get the computed status of a job
+create definer = 'select_objects'@'localhost'
+  function job_status_func (jobid binary(16))
+    returns varchar(32)
+    comment 'Get the status of a job'
+    reads sql data sql security definer
+  begin
+    declare status varchar(32);
+    declare allnamed boolean;
+    set status = (select jobs.status from jobs where id = jobid);
+
+    if status = 'created' then
+      set allnamed = (
+        select bit_and(present) from job_data where job_id = jobid
+      ) = 1;
+      if allnamed then
+        return 'prepared';
+      else
+        return 'incomplete';
+      end if;
+    end if;
+    return status;
+  end;
```

Comment:
Do we need a "checkout job" type procedure for marking the job as "running" when a worker process picks it up to compute? I'm not sure if that'd be something we compute on the fly like we're doing here, but it might be an additional condition for disallowing `add_job_data` and `queue_job`.

Reply:
My thinking was that the actual status that gets returned from the API will query both the db and the queueing system. I think we probably first check the db and if `status =='queued'`, see if the job is actually running to return 'running'. Otherwise, return the db status. This way there is no state where the job started, updated the db with 'running', and then failed in some way but never updated the db. Maybe a helper script/cron is needed to check for queued jobs that have failed running for some reason but haven't updated their status to 'error'. 

I probably should have renamed `check_job_queued`, but it checks if status is anything after `created` in the status enum so `add_job_data` fails appropriately.

Reply:
Cool that makes sense, less state management to handle. :+1: